server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java

```diff
@@ -0,0 +1,252 @@
+        try {
+            return this.openMultipartStreams(partSize);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private StreamContext openMultipartStreams(long partSize) throws IOException {
+        if (inputStreams.get() != null) {
+            throw new IOException("Multi-part streams are already created.");
+        }
+
+        this.partSize = partSize;
+        this.lastPartSize = (contentLength % partSize) != 0 ? contentLength % partSize : partSize;
+        this.numberOfParts = (int) ((contentLength % partSize) == 0 ? contentLength / partSize : (contentLength / partSize) + 1);
+        InputStream[] streams = new InputStream[numberOfParts];
+        inputStreams.set(streams);
+
+        return new StreamContext(getTransferPartStreamSupplier(), partSize, lastPartSize, numberOfParts);
+    }
+
+    private CheckedTriFunction<Integer, Long, Long, InputStreamContainer, IOException> getTransferPartStreamSupplier() {
+        return ((partNo, size, position) -> {
+            assert inputStreams.get() != null : "expected inputStreams to be initialised";
+            return getMultipartStreamSupplier(partNo, size, position).get();
+        });
+    }
+
+    /**
+     * OffsetRangeInputStreamSupplier is used to get the offset based input streams at runtime
+     */
+    public interface OffsetRangeInputStreamSupplier {
+        OffsetRangeInputStream get(long size, long position) throws IOException;
+    }
+
+    interface LocalStreamSupplier<Stream> {
+        Stream get() throws IOException;
+    }
+
+    private LocalStreamSupplier<InputStreamContainer> getMultipartStreamSupplier(
+        final int streamIdx,
+        final long size,
+        final long position
+    ) {
+        return () -> {
+            try {
+                OffsetRangeInputStream offsetRangeInputStream = offsetRangeInputStreamSupplier.get(size, position);
+                InputStream inputStream = !isRemoteDataIntegrityCheckPossible()
+                    ? new ResettableCheckedInputStream(offsetRangeInputStream, fileName)
+                    : offsetRangeInputStream;
+                Objects.requireNonNull(inputStreams.get())[streamIdx] = inputStream;
+
+                return new InputStreamContainer(inputStream, size, position);
+            } catch (IOException e) {
+                log.error("Failed to create input stream", e);
+                throw e;
+            }
+        };
+    }
+
+    private boolean isRemoteDataIntegrityCheckPossible() {
+        return isRemoteDataIntegritySupported && !areInputStreamsDecorated;
+    }
+
+    private void finalizeUpload(boolean uploadSuccessful) {
+        if (isRemoteDataIntegrityCheckPossible()) {
+            return;
+        }
+
+        if (uploadSuccessful) {
+            long actualChecksum = getActualChecksum();
+            if (actualChecksum != expectedChecksum) {
+                throw new RuntimeException(
```

Comment:
Why wrap it with a RunTime Exception?

Reply:
Either this or we extend the UploadFinalizer interface with exception and handle the checked exception in the future explicitly in s3 plugin.

Reply:
how does this relates to handling in plugin ? This will come in picture when plugin doesn't do integrity check . 

Extending `UploadFinalizer interface with exception` looks to better approach IMO .

Reply:
how does this relates to handling in plugin ? This will come in picture when plugin doesn't do integrity check . 

Extending `UploadFinalizer interface with exception` looks to better approach IMO .

Reply:
Resolving this for now based on discussion with @gbbafna.